// @subql/common/dist/project/versioned/ProjectManifestVersioned.js
"use strict";
// Copyright 2020-2021 OnFinality Limited authors & contributors
// SPDX-License-Identifier: Apache-2.0
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProjectManifestVersioned = exports.manifestIsV0_2_0 = exports.manifestIsV0_0_1 = void 0;
const class_transformer_1 = require("class-transformer");
const v0_0_1_1 = require("./v0_0_1");
const v0_2_0_1 = require("./v0_2_0");
const SUPPORTED_VERSIONS = {
    '0.0.1': v0_0_1_1.ProjectManifestV0_0_1Impl,
    '0.2.0': v0_2_0_1.ProjectManifestV0_2_0Impl,
};
function manifestIsV0_0_1(manifest) {
    return manifest.specVersion === '0.0.1';
}
exports.manifestIsV0_0_1 = manifestIsV0_0_1;
function manifestIsV0_2_0(manifest) {
    return manifest.specVersion === '0.2.0';
}
exports.manifestIsV0_2_0 = manifestIsV0_2_0;
class ProjectManifestVersioned {
    constructor(projectManifest) {
        const klass = SUPPORTED_VERSIONS[projectManifest.specVersion];
        if (!klass) {
            throw new Error('specVersion not supported for project manifest file');
        }
        this._impl = (0, class_transformer_1.plainToClass)(klass, projectManifest);
    }
    get asImpl() {
        return this._impl;
    }
    get isV0_0_1() {
        return this.specVersion === '0.0.1';
    }
    get asV0_0_1() {
        return this._impl;
    }
    get isV0_2_0() {
        return this.specVersion === '0.2.0';
    }
    get asV0_2_0() {
        return this._impl;
    }
    toDeployment() {
        return this.toDeployment();
    }
    validate() {
        this._impl.validate();
    }
    get dataSources() {
        return this._impl.dataSources;
    }
    get schema() {
        if (manifestIsV0_0_1(this._impl)) {
            return this._impl.schema;
        }
        return this._impl.schema.file;
    }
    get specVersion() {
        return this._impl.specVersion;
    }
    get description() {
        return this._impl.description;
    }
    get repository() {
        return this._impl.repository;
    }
}
exports.ProjectManifestVersioned = ProjectManifestVersioned;
