// @polkadot/types/codec/WrapperOpaque.cjs
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.WrapperOpaque = void 0;

var _util = require("@polkadot/util");

var _Base = require("./Base.cjs");

var _index = require("./utils/index.cjs");

// Copyright 2017-2021 @polkadot/types authors & contributors
// SPDX-License-Identifier: Apache-2.0
function decodeRaw(registry, Type, value) {
  const Clazz = (0, _index.typeToConstructor)(registry, Type);

  if ((0, _util.isU8a)(value)) {
    const [, u8a] = (0, _util.compactStripLength)(value);
    return new Clazz(registry, u8a);
  }

  return new Clazz(registry, value);
}

class WrapperOpaque extends _Base.Base {
  constructor(registry, Type, value) {
    super(registry, decodeRaw(registry, Type, value));
  }

  static with(Type) {
    return class extends WrapperOpaque {
      constructor(registry, value) {
        super(registry, Type, value);
      }

    };
  }
  /**
   * @description Returns the base runtime type name for this instance
   */


  toRawType() {
    return `WrapperOpaque<${this._raw.toRawType()}>`;
  }
  /**
   * @description The length of the value when encoded as a Uint8Array
   */


  toU8a(isBare) {
    const u8a = super.toU8a(isBare);
    return isBare ? u8a : (0, _util.compactAddLength)(u8a);
  }
  /**
   * @description Return the inner value for the wrapped type
   */


  unwrap() {
    return this._raw;
  }

}

exports.WrapperOpaque = WrapperOpaque;
