// @subql/common/dist/project/versioned/v0_2_0/model.js
"use strict";
// Copyright 2020-2021 OnFinality Limited authors & contributors
// SPDX-License-Identifier: Apache-2.0
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProjectManifestV0_2_0Impl = exports.DeploymentV0_2_0 = exports.CustomDataSourceV0_2_0Impl = exports.RuntimeDataSourceV0_2_0Impl = exports.ProjectMappingV0_2_0 = exports.ProjectNetworkV0_2_0 = exports.ProjectNetworkDeploymentV0_2_0 = exports.FileType = void 0;
const class_transformer_1 = require("class-transformer");
const class_validator_1 = require("class-validator");
const js_yaml_1 = __importDefault(require("js-yaml"));
const models_1 = require("../../models");
const base_1 = require("../base");
class FileType {
}
__decorate([
    (0, class_validator_1.IsString)(),
    __metadata("design:type", String)
], FileType.prototype, "file", void 0);
exports.FileType = FileType;
class ProjectNetworkDeploymentV0_2_0 {
}
__decorate([
    (0, class_validator_1.IsString)(),
    __metadata("design:type", String)
], ProjectNetworkDeploymentV0_2_0.prototype, "genesisHash", void 0);
__decorate([
    (0, class_validator_1.ValidateNested)(),
    (0, class_transformer_1.Type)(() => FileType),
    (0, class_validator_1.IsOptional)(),
    __metadata("design:type", FileType)
], ProjectNetworkDeploymentV0_2_0.prototype, "chaintypes", void 0);
exports.ProjectNetworkDeploymentV0_2_0 = ProjectNetworkDeploymentV0_2_0;
class ProjectNetworkV0_2_0 extends ProjectNetworkDeploymentV0_2_0 {
}
__decorate([
    (0, class_validator_1.IsString)(),
    (0, class_validator_1.IsOptional)(),
    __metadata("design:type", String)
], ProjectNetworkV0_2_0.prototype, "endpoint", void 0);
__decorate([
    (0, class_validator_1.IsString)(),
    (0, class_validator_1.IsOptional)(),
    __metadata("design:type", String)
], ProjectNetworkV0_2_0.prototype, "dictionary", void 0);
exports.ProjectNetworkV0_2_0 = ProjectNetworkV0_2_0;
class ProjectMappingV0_2_0 extends models_1.Mapping {
}
__decorate([
    (0, class_validator_1.IsString)(),
    __metadata("design:type", String)
], ProjectMappingV0_2_0.prototype, "file", void 0);
exports.ProjectMappingV0_2_0 = ProjectMappingV0_2_0;
class RuntimeDataSourceV0_2_0Impl extends models_1.RuntimeDataSourceBase {
}
__decorate([
    (0, class_transformer_1.Type)(() => ProjectMappingV0_2_0),
    (0, class_validator_1.ValidateNested)(),
    __metadata("design:type", Object)
], RuntimeDataSourceV0_2_0Impl.prototype, "mapping", void 0);
exports.RuntimeDataSourceV0_2_0Impl = RuntimeDataSourceV0_2_0Impl;
class CustomDataSourceV0_2_0Impl extends models_1.CustomDataSourceBase {
}
exports.CustomDataSourceV0_2_0Impl = CustomDataSourceV0_2_0Impl;
class DeploymentV0_2_0 {
}
__decorate([
    (0, class_validator_1.Equals)('0.2.0'),
    (0, class_validator_1.IsString)(),
    __metadata("design:type", String)
], DeploymentV0_2_0.prototype, "specVersion", void 0);
__decorate([
    (0, class_validator_1.ValidateNested)(),
    (0, class_transformer_1.Type)(() => FileType),
    __metadata("design:type", FileType)
], DeploymentV0_2_0.prototype, "schema", void 0);
__decorate([
    (0, class_validator_1.IsArray)(),
    (0, class_validator_1.ValidateNested)(),
    (0, class_transformer_1.Type)(() => CustomDataSourceV0_2_0Impl, {
        discriminator: {
            property: 'kind',
            subTypes: [{ value: RuntimeDataSourceV0_2_0Impl, name: 'substrate/Runtime' }],
        },
        keepDiscriminatorProperty: true,
    }),
    __metadata("design:type", Array)
], DeploymentV0_2_0.prototype, "dataSources", void 0);
__decorate([
    (0, class_validator_1.ValidateNested)(),
    (0, class_transformer_1.Type)(() => ProjectNetworkDeploymentV0_2_0),
    __metadata("design:type", ProjectNetworkDeploymentV0_2_0)
], DeploymentV0_2_0.prototype, "network", void 0);
exports.DeploymentV0_2_0 = DeploymentV0_2_0;
class ProjectManifestV0_2_0Impl extends base_1.ProjectManifestBaseImpl {
    toDeployment() {
        return js_yaml_1.default.dump(this._deployment, {
            sortKeys: true,
            condenseFlow: true,
        });
    }
    get deployment() {
        if (!this._deployment) {
            this._deployment = (0, class_transformer_1.plainToClass)(DeploymentV0_2_0, this);
            (0, class_validator_1.validateSync)(this._deployment, { whitelist: true });
        }
        return this._deployment;
    }
    validate() {
        const errors = (0, class_validator_1.validateSync)(this.deployment, { whitelist: true, forbidNonWhitelisted: true });
        if (errors === null || errors === void 0 ? void 0 : errors.length) {
            // TODO: print error details
            const errorMsgs = errors.map((e) => e.toString()).join('\n');
            throw new Error(`failed to parse project.yaml.\n${errorMsgs}`);
        }
    }
}
__decorate([
    (0, class_validator_1.Equals)('0.2.0'),
    __metadata("design:type", String)
], ProjectManifestV0_2_0Impl.prototype, "specVersion", void 0);
__decorate([
    (0, class_validator_1.IsString)(),
    __metadata("design:type", String)
], ProjectManifestV0_2_0Impl.prototype, "name", void 0);
__decorate([
    (0, class_validator_1.IsString)(),
    __metadata("design:type", String)
], ProjectManifestV0_2_0Impl.prototype, "version", void 0);
__decorate([
    (0, class_validator_1.IsObject)(),
    (0, class_validator_1.ValidateNested)(),
    (0, class_transformer_1.Type)(() => ProjectNetworkV0_2_0),
    __metadata("design:type", ProjectNetworkV0_2_0)
], ProjectManifestV0_2_0Impl.prototype, "network", void 0);
__decorate([
    (0, class_validator_1.ValidateNested)(),
    (0, class_transformer_1.Type)(() => FileType),
    __metadata("design:type", FileType)
], ProjectManifestV0_2_0Impl.prototype, "schema", void 0);
__decorate([
    (0, class_validator_1.IsArray)(),
    (0, class_validator_1.ValidateNested)(),
    (0, class_transformer_1.Type)(() => CustomDataSourceV0_2_0Impl, {
        discriminator: {
            property: 'kind',
            subTypes: [{ value: RuntimeDataSourceV0_2_0Impl, name: 'substrate/Runtime' }],
        },
        keepDiscriminatorProperty: true,
    }),
    __metadata("design:type", Array)
], ProjectManifestV0_2_0Impl.prototype, "dataSources", void 0);
exports.ProjectManifestV0_2_0Impl = ProjectManifestV0_2_0Impl;
